### PR TITLE
Enhance configuration API

### PR DESCRIPTION
**Pull request by sbstjn**
* `table` property supports string and array as value(s)
* Support for `index` property to scale `GlobalSecondaryIndexes` #2 

### Tasks

* [x] Rewrite index support
* [x] Rewrite `DependsOn` handling
* [x] Update `README.md` 
* [x] Add info about [CloudWatch API rate limits](http://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_limits.html)
* [x] Add stage to naming schema
* [x] Extend tests for `names.js`

This branch supports flexible Auto Scaling for tables and/or global secondary indexes with the following configurations:

```yaml
custom:
  capacities:
    - table: CustomTableA
      index: custom-index-global
      indexOnly: true
      read:
        minimum: 1
        maximum: 123
        usage: 0.80
      write:
        minimum: 10
        maximum: 100
        usage: 0.50
    - table: CustomTableB
      read:
        minimum: 1
        maximum: 123
        usage: 0.80
      write:
        minimum: 10
        maximum: 100
        usage: 0.50
    - table:
        - CustomTableC
        - CustomTableD
        - CustomTableE
      index:
        - custom-index
      read:
        minimum: 1
        maximum: 123
        usage: 0.80
      write:
        minimum: 10
        maximum: 100
        usage: 0.50
```

### CODE REVIEW

**Comment by charleswhchan** [comment]
1. Is `indexOnly` applicable for the "string style" only? (I am guessing not, but just want to be clear) To eliminate the need for this parameter, is it possibly to simply provide the index name as the resource without table name?
2. The array format looks interesting and useful, but may become confusing when there are lots of tables and indexes. I supposed as long as your indexes follows some form of consistent naming format wrt to the table name it should't lead to readability problems.

Thanks @sbstjn :+1:

**Comment by sbstjn** [comment]
Thanks for the feedback. Based on the CloudFormation resource I think the global secondary index has no CloudFormation resource identifier. They seem to be always referenced by the combination of the table resource/name and the name of the index:

```yaml
    CustomTableC:
      Type: AWS::DynamoDB::Table
      Properties:
        TableName: custom-table-C
        GlobalSecondaryIndexes:
          - IndexName: custom-index
            KeySchema:
              - AttributeName: key
                KeyType: HASH
            Projection:
              ProjectionType: ALL
            ProvisionedThroughput:
              ReadCapacityUnits: 5
              WriteCapacityUnits: 5
```

Based on this, I think it's not possible to just provide an `index` without `table`.

The `indexOnly` property should be used if you only want to apply autoscaling on the index, and not on the table. I don't know if this is real use case ;)

**Comment by sbstjn** [comment]
I have extended the `README.md` with information about the `indexOnly` parameter. The list-support is working and tested, but to avoid irritation I'll skip explaining the feature in the readme ;)

File: README.md [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
 [![license](https://img.shields.io/github/license/sbstjn/serverless-dynamodb-autoscaling.svg)](https://github.com/sbstjn/serverless-dynamodb-autoscaling/blob/master/LICENSE.md)
 [![Coveralls](https://img.shields.io/coveralls/sbstjn/serverless-dynamodb-autoscaling.svg)](https://coveralls.io/github/sbstjn/serverless-dynamodb-autoscaling)
 
-With this plugin for [serverless](https://serverless.com) you can set DynamoDB Auto Scaling configuratin in your `serverless.yml` file. The plugin supports multiple tables and separate configuration sets for `read` and `write` capacities using AWS [native DynamoDB Auto Scaling](https://aws.amazon.com/blogs/aws/new-auto-scaling-for-amazon-dynamodb/).
+With this plugin for [serverless](https://serverless.com), you can enable DynamoDB Auto Scaling for tables and **Global Secondary Indexes** easily in your `serverless.yml` configuration file. The plugin supports multiple tables and indexes, as well separate configuration sets for `read` and `write` capacities using Amazon's [native DynamoDB Auto Scaling](https://aws.amazon.com/blogs/aws/new-auto-scaling-for-amazon-dynamodb/).
```
Comment (charleswhchan) [inline README.md]: ... as well **as** separate configuration ~~sets~~ for `read` and `write`

File: src/plugin.js [hunk 2]
```diff
@@ -9,26 +7,42 @@ const Target = require('./aws/target')
...
     assert(this.serverless.service, 'Invalid serverless configuration')
     assert(this.serverless.service.provider, 'Invalid serverless configuration')
     assert(this.serverless.service.provider.name, 'Invalid serverless configuration')
     assert(this.serverless.service.provider.name === 'aws', 'Only supported for AWS provider')
 
-    assert(this.serverless.service.custom.capacities, 'No Auto Scaling configuration found')
+    assert(this.serverless.service.provider.stage, 'Invalid serverless configuration')
+
+    assert(this.serverless.service.custom, 'Not Auto Scaling configuration found')
+    assert(this.serverless.service.custom.capacities, 'Not Auto Scaling configuration found')
```
Comment (charleswhchan) [inline src/plugin.js]: No~~t~~ Auto Scaling configuration found

**Comment by sbstjn** [comment]
👍

File: src/plugin.js [hunk 3]
```diff
@@ -42,54 +56,123 @@ class Plugin {
...
-          resources.push(
-            new Policy(table.table, table.write.usage, false, 60, 60),
-            new Target(table.table, table.write.minimum, table.write.maximum, false)
-          )
-        }
+  /**
+   * Generate CloudFormation resources for DynamoDB table and indexes
+   *
+   * @param {string} table
+   * @param {obejct} config
```
Comment (charleswhchan) [inline src/plugin.js]: typo: object

**Comment by charleswhchan** [comment]
Thanks, @sbstjn. Except for a couple of typos, changes looks good :+1:

**Comment by sbstjn** [comment]
Thanks for checking 👍

**Comment by nasirumair** [comment]
@sbstjn Is there any support for scaling tables in other stacks at all? I am trying to autoscale a table that is currently existing but not in the stack

**Pull request merged**